### PR TITLE
Stylize tsc's error messages with color and context

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This change seeks to make our output a little easier on the eyes and addresses #3852. I suggest this PR's final changes be reviewed rather than going commit-by-commit.

Firstly, a new flag is introduced: `--diagnosticStyle`. It accepts the options `pretty` and `simple`.  It is currently marked as experimental, and we can bikeshed over the names later.

If `--diagnosticStyle simple` is used, the user will get the same behavior as they experience today. **Note that `simple` is left as the default.**

If `--diagnosticStyle pretty` is used as an option, then we check if we are outputting to a terminal-style descriptor. If we are *entirely sure*, then we stylize the output with color. If not (e.g. output is being piped, forwarded to a file, etc.) we make the output readable so that no special control characters are used in the output.

Here's some screenshots/gifs to give you an idea of what this looks like right now.

watchpretty

Monokai
image

Solarized Light
image

Terminal Basic
image

My Stupid Terminal Theme
image

Cygwin's MinTTY
image

cmd.exe
image

GNOME Terminal Defaults in Ubuntu
image

### CODE REVIEW

**Comment by myitcv** [comment]
> Note that simple is left as the default.

Thanks. This should mean that any tools that consume the output from `tsc` continue to work.

**Comment by DanielRosenwasser** [comment]
@myitcv this was one of the original concerns I heard a lot of from other members of the team; however, I think it would be a good idea to make it a default option in `tsconfig.json` when using `tsc --init`.

**Comment by Gaelan** [comment]
One thing I would really appreciate would be some sort of coloring for the "starting incremental recompilation" message, so we can more easily tell which errors are current.

**Comment by tinganho** [comment]
Awesome that it shows the line numbers! I would also like the message when overflowed to be tabbed to the starting position. For example, the text part `type 'I'` was too long and began on the next line on column 0. It would be awesome if that part began in column 2 instead. 

I guess you would add colors on the messages too?

**Comment by DanielRosenwasser** [comment]
@tinganho I think terminal size detection/wrapping is better left for a separate issue/PR.

File: src/compiler/tsc.ts [hunk 1]
```diff
@@ -95,10 +104,78 @@ namespace ts {
...
+        if (diagnostic.file) {
+            let { start, length, file } = diagnostic;
+            let { line: firstLine, character: firstLineChar } = getLineAndCharacterOfPosition(file, start);
+            let { line: lastLine, character: lastLineChar } = getLineAndCharacterOfPosition(file, start + length);
+            const lastLineInFile = getLineAndCharacterOfPosition(file, file.text.length).line;
+
+            let hasMoreThanFiveLines = (lastLine - firstLine) >= 4;
+            let gutterWidth = (lastLine + 1 + "").length;
+            if (hasMoreThanFiveLines) {
+                gutterWidth = Math.max("...".length, gutterWidth);
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: capture the ellipses in a const, and so is its length.

Comment (mhegazy) [inline src/compiler/tsc.ts]: also applies to the squiggle down below.

File: src/compiler/tsc.ts [hunk 2]
```diff
@@ -95,10 +104,78 @@ namespace ts {
...
+                if (i === firstLine) {
+                    // If we're on the last line, then limit it to the last character of the last line.
+                    // Otherwise, we'll just squiggle the rest of the line, giving 'slice' no end position.
+                    let lastCharForLine = i === lastLine ? lastLineChar : undefined;
+
+                    output += lineContent.slice(0, firstLineChar).replace(/\S/g, " ");
+                    output += lineContent.slice(firstLineChar, lastCharForLine).replace(/./g, "~");
+                }
+                else if (i === lastLine) {
+                    output += lineContent.slice(0, lastLineChar).replace(/./g, "~");
```
Comment (mhegazy) [inline src/compiler/tsc.ts]: sound like a long way to say `repeat("~", lastLineChar)`, where repeat is just a simple for loop.

**Comment by mhegazy** [comment]
LGTM. a few nits, but other than than that :+1:

**Comment by mhegazy** [comment]
please add a snippit about it  in https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript

**Pull request merged**